File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
    int rows = grid.size();
    int cols = grid[0].size();
    int islandCount = 0;

    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            if (grid[i][j] == '1') {
                bfs(grid, i, j);
                islandCount++;
            }
        }
    }

    return islandCount;
    }

    void bfs(vector<vector<char>>& grid, int x, int y) {
    int rows = grid.size();
    int cols = grid[0].size();
    queue<pair<int, int>> q;
    q.push({x, y});
    grid[x][y] = '0';
    int directions[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}}; 

    while (!q.empty()) {
        auto [cx, cy] = q.front();
        q.pop();

        for (auto& dir : directions) {
            int nx = cx + dir[0];
            int ny = cy + dir[1];

            if (nx >= 0 && nx < rows && ny >= 0 && ny < cols && grid[nx][ny] == '1') {
                q.push({nx, ny});
                grid[nx][ny] = '0';
            }
        }
    }
}
};

File: 200-number-of-islands/number-of-islands.ts
function numIslands(grid: string[][]): number {
    if (grid.length == 0) return 0;
    let numIslands: number = 0;
    const depthSearch = (grid: string[][] ,i: number, j: number) => {
        if ( ( i < 0 ) || ( i >= grid.length )  || ( j < 0 )  || ( j >= grid[i].length ) || ( grid[i][j] !== "1") ) return
        grid[i][j] = "0";
        depthSearch(grid,i+1,j);
        depthSearch(grid,i,j+1);
        depthSearch(grid,i-1,j);
        depthSearch(grid,i,j-1);
    }
    for (let i = 0; i < grid.length; i++) {
        for (let j = 0; j < grid[i].length; j++) {
            if (grid[i][j] === "1"){
                numIslands++;
                depthSearch(grid,i,j);
                }
            }
        }

    return numIslands;
};
